docs/parser/ast-generation.md

```diff
@@ -0,0 +1,234 @@
+                "(Ljava/lang/Object;Lorg/enso/ast/Ast$Shape;)V"
+            ),
+        }
+    }
+}
+```
+
+This struct exposes a low level type-unsafe API for instantiating the Scala AST.
+As an example the Scala`Ast.Num.Int(10)` would be constructed by
+`Scala::new(&env).int.init(&[10.into()])`. However, use of this API is
+discoureged. Instead, you should use the high-level
+[Constructor API](#constructor-api).
+
+## Constructor API
+
+This is a high level constructor API autogenerated from the Rust AST definition.
+The interface is defined by the trait `ast::api::Api` and implemented by the
+structs `ast::api::Scala` and `ast::api::Rust`. It presents a uniform way to
+instantiate both the Rust and Scala AST by a single function:
+
+```rust
+use ast::api;
+use jni::JNIEnv;
+
+pub fn build<Api:api::Api>(api:&Api) -> Api::AstShape { .. }
+
+pub fn main(env:JNIEnv) {
+    let scala_ast = build(api::Scala::new(&env));
+    let rust_ast  = build(api::Rust);
+}
+```
+
+The API tries to follow the AST constructors as close to possible, but some
+modifications had to be made due to the trait limitations:
+
+- All types are moved into the same scope.
+- Types parameters are monomorphized. (This won't be necessary with GAT.)
+
+To give you a sense what that means:
+
+```rust
+pub enum Shape { Num(num::Number), App(app::App<Shape>) }
+
+pub mod num {
+    pub enum Number { Int{num:i64}, Real{num:f64} }
+}
+pub mod app {
+    pub struct App<T>{fun:Box<T>, arg:Box<Shape> }
+}
+```
+
+Becomes:
+
+```rust
+pub trait Api {
+    type Shape;
+    type Number;
+    /// The AST Node app::App<Shape>.
+    type AppShape;
+
+    fn shape_num   (&self, variant:Self::Number)        -> Self::Shape;
+    fn shape_app   (&self, variant:Self::App)           -> Self::Shape;
+    fn number_int  (&self, num:i64)                     -> Self::Number;
+    fn number_real (&self, num:f64)                     -> Self::Number;
+    fn app_shape
+    (&self, fun:Box<Self::Shape>, arg:Box<Self::Shape>) -> Self::AppShape;
+}
+```
+
+The precise rules for structs are following:
+
+```
+struct A(i32)                => type A; fn a(&self, val0:i32) -> Self::A;
+struct B<T>{t:T}             => .. no output ..
+struct C<T>{t:T, b:B<A>, ..} => type BA; fn ba(&self, t:Self::A) -> Self::BA;
+```
+
+The rules for enums are the same, except one type can have multiple
+constructors:
+
+```
+enum Foo { A{..}, B{..} } => {
+    type Foo;
+    fn foo_a(&self, ..) -> Self::Foo;
+    fn foo_b(&self, ..) -> Self::Foo;
+}
+```
+
+Note that builtin types (types that are not defined in `ast.rs`) are left intact 
+(they do not introduce an additional associated type). Therefore, they need
+to implement the trait `ast::ffi::ToJValue` to enable the autogenerated
+[Scala Bindings](#scala-bindings) to convert those types to correct JNI values.
+
+## Performance
+
+On benchmarks that allocate 2e6 AST nodes the Rust API takes ~0.4s while the 
+Scala API takes ~4s. 
```

Comment:
You need to describe the benchmarking and profiling approaches taken here.

Additionally, you should test instantiation of something small in a loop via JNI and check the overhead. I would test via `jni-rs`, but also via the c bindings directly.